File: 947-most-stones-removed-with-same-row-or-column/947-most-stones-removed-with-same-row-or-column.cpp
class Solution {
public:
    vector<int> parent;
    int findPar(int node){
        if(node==parent[node]) return node;
        return parent[node] = findPar(parent[node]);
    }
    
    bool unionn(int u,int v){
        u = findPar(u);
        v = findPar(v);
        
        if(u==v) return false;
        parent[v]=u;
        return true;
    }
    int removeStones(vector<vector<int>>& stones) {
        int n = stones.size();
        parent.resize(n);
        
        for(int i=0;i<n;i++){
            parent[i]=i;
        }
        int cnt=0;
        for(int i=0;i<n;i++){
            for(int j=i+1;j<n;j++){
                if(stones[i][0]==stones[j][0] || (stones[i][1]==stones[j][1]))
                    if(unionn(i,j))
                        cnt++;
            }
        }
        return cnt;
    }
};

File: 947-most-stones-removed-with-same-row-or-column/947-most-stones-removed-with-same-row-or-column.java
class DSU 
{
    int parent[];
    int height[];
 
    public DSU(int n)
    {
        parent = new int[n];
        height = new int[n];
        Arrays.fill(parent , -1);
        Arrays.fill(height , 1);
    }
 
    public int findRoot(int node)
    {
        if(parent[node] == -1)
            return node;
        return (parent[node] = findRoot(parent[node]));
    }
 
    public void union(int node1 , int node2)
    {
        int ra = findRoot(node1);
        int rb = findRoot(node2);
 
        if(ra != rb)
        {
            if(height[ra] < height[rb])
                parent[ra] = rb;
            else if(height[rb] < height[ra])
                parent[rb] = ra;
            else 
            {
                parent[rb] = ra;
                height[ra] ++;
            }
        }
    }
}
class Solution {
    public int removeStones(int[][] stones) {
        DSU dsu = new DSU(20001);
        for(int stone[]:stones){
            dsu.union(stone[0],stone[1]+10001);
        }
        Set<Integer> set = new HashSet<>();
        for(int stone[]:stones){
            set.add(dsu.findRoot(stone[0]));
        }
        int comps=set.size();
        return stones.length-comps;
    }
}